runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -3917,6 +3918,118 @@ J9::CodeGenerator::collectSymRefs(
+         {
+         TR_VirtualGuard *vg = comp->findVirtualGuardInfo(node);
+         // Mainly we need to make sure that virtual guard which performs the TR_MethodTest and can be NOP'd are needed the range check.
+         if (vg && vg->getTestType() == TR_MethodTest && 
+            !(comp->performVirtualGuardNOPing() && (node->isNopableInlineGuard() || comp->isVirtualGuardNOPingRequired(vg))))
+            {
+            TR::SymbolReference *callSymRef = vg->getSymbolReference();
+            TR_ASSERT_FATAL(callSymRef != NULL, "Guard n%dn for the inlined call should have stored symbol reference for the call", node->getGlobalIndex());
+            // TODO: If node is nop'd then we do not need this range check. 
+            if (callSymRef->getSymbol()->castToMethodSymbol()->isInterface())
+               {
+               TR::DebugCounter::incStaticDebugCounter(comp, TR::DebugCounter::debugCounterName(comp, "profiledInterfaceTest/({%s}{%s})", comp->signature(), comp->getHotnessName(comp->getMethodHotness())));
+               dumpOptDetails(comp, "Need to add a rangecheck before n%dn in block_%d\n",node->getGlobalIndex(), block->getNumber());
+
+               // We need a VFT Load of the receiver object to get the VTableHeader.size to check the range. As this operation is happening during codegen phase, only
+               // known concrete way we can have this information is through aloadi child of the guard that has single child which is vft load of receiver object.
+               // Now instead of accessing VFT load from the child of the aloadi, we could have treetop's the aloadi during inlining where we generate the virtual guard
+               // to access information from the treetop. Because of the same reasons lined up behind adding range check test during codegen phase in the description of this function,
+               // we would need to make changes in all optimizations moving Virtual Guard around to keep that treetop together before guard which will be very difficult to enforce.
+               // Also as children of virtual guard is very self contained and atm it is very unlikely that other optimizations are going to find opportunity of manipulating them and
+               // Because of the fact that it is very unlikely that we will have another aloadi node with same VTable offset of same receiver object, this child would not be commoned out
+               // and have only single reference in this virtual guard therefore splitting of block will not store it to temp slot.
+               // Due to the above reasons we have a fatal assert in case the first child of the guard node performing is not aloadi. If there exists any case we need to reconsider our choice for range check test.
+               TR_ASSERT_FATAL(node->getFirstChild()->getOpCodeValue() == TR::aloadi, "Virtual Guard perfoming TR_MethodTest should have aloadi as first child, to extract the vftPointer of the receiver object");
```

Comment:
@andrewcraik  @jdmpapin  I have added a comment here regarding to why we have the FATAL_ASSERT if child of the guard is not `aloadi` .

Reply:
I've recently seen PRE hoist an `aloadi <vft-symbol>` out of a loop, leaving a profiled guard with a VFT test of this form:
```
ifacmpne --> slow path
  aload <temp slot ...>
  aconst 0x... (isClassPointerConstant)
```
There might be circumstances where `aloadi <vtable-entry-symbol>` looks like it will happen unconditionally, so that normally it would be safe to hoist, but because it's part of one of these guards it would now be implicitly conditional.

Reply:
@andrewcraik  I discussed this with @jdmpapin  as well, although it is possible to optimization to hoist out  `aloadi` we would still left with the issue where that `aloadi` is not safe as it is followed by invisible condition.  Another solution Devin came up with was , we could have new type of conditional load node in trees which we can identify during codegen phase and generate a range check test. Even in this case, if some optimization hoist id out of loop, we would have difficulty generating a range check test. 
It seems if some optimization tries to work with these type of load, it needs to be cautious not to mess up these kind of invisible conditional node. 
@andrewcraik  I would like to get your view on this. ATM I have strengthen out the ASSERT by also checking the type of symref associated with `aloadi` which should be of type vTable Entry Symbol in https://github.com/eclipse/openj9/pull/4721/commits/ddbed9715f1f2fd5b5f4b123bbbe809d09bee454